Rule Processing: pass in default

psealock opened the pull request:
Pass in `$default` value to `dotNotation` rule. The value wasn't passed down to the `transform` function and wasn't being returned when a path didn't exist in an object.

psealock commented:
@moon0326 Would you have a moment to take a look at this? If the direction is Ok, I will add unit tests and make it ready for review.

joshuatf reviewed (commented):
I'll let Moon weigh in here, but I think this works well for this use case.

It would also be really cool to have a `.length` property as I think this could prove useful in other areas.  Guessing this would need to be checked in `DotNotation::get()` and be a reserved keyword (can't be used as an index).

moon0326 commented:
@psealock Thank you for catching it. I should have considered supporting the default value. This looks good to me 👍 

I like @joshuatf suggestion to have a way to count the array. I think we can add `array_count`, but the default value should work as well.

psealock commented:
@moon0326  this one is ready for review. I fixed failing tests and added a test case for the default value. I will tackle `array_count` in a follow up.

moon0326 reviewed (approved):
LGTM and tests passed as expected. Nice work 👍 🎉 🚢
